02-intermediate-front-end/events-and-callbacks/readme.md

```diff
@@ -33,77 +76,592 @@ button.addEventListener("click", handleClickEvent)
+  - Why?
+  - How is this different from events?
+  - When you do this, why doesn't the "stop" button seem to work?
+- How is `clearInterval` different from `removeEventListener`?
+- Give `singAnnoyingSong` an argument of `e`, like we did for the event listeners. What information does it contain?
+
+## Making your own callbacks
+
+We've done a lot of stuff with callbacks today, passing them as arguments into built-in Javascript functions. But we haven't tried passing them into functions we wrote ourselves.
+
+We're going to create a script that shows your interal monologue over the course of a normal weekday. Create three functions: one that `console.logs` something you think to yourself on waking up, another something you think during your commute, and that last something you think when you get to GA.
+
+Then, create a function called `averageMorning` that takes three arguments, `first`, `then`, and `lastly`.
+
+These three arguments represent three functions. Inside `averageMorning`, call `first`, `then`, and `lastly`, in order.
+
+Then, call the `averageMorning` function itself, passing in as arguments the three functions you wrote.
+
+The result should be something like this:
+
+```js
+function haveBreakfast(){
+  console.log("Gotta have my Pops!")
+}
+function commute(){
+  console.log("What an interesting smell on Metro this morning...")
+}
+function workAtGA(){
+  console.log("At work! Now I can promote RobertAKARobin.com through the disguise of teaching again.")
+}
+function averageMorning(first, then, lastly){
+  console.log("This is my internal monologue on a normal weekday")
+  first();
+  then();
+  setTimeout(lastly, 5000);
+}
+averageMorning(haveBreakfast, commute, workAtGA);
+```
+
+## IIFEs
+
+Want to see something really crazy? We can define *and* call the `averageMorning` function all at once, without needing `averageMorning()` on a separate line. Just delete a couple things, and put in two extra parens:
+
+```js
+(function averageMorning(first, then, lastly){
+  console.log("This is my internal monologue on a normal weekday")
+  first();
+  then();
+  setTimeout(lastly, 5000);
+}(haveBreakfast, commute, workAtGA));
+```
+
+This is called an **Immediately-Invoked Function Expression**, which is a fancy way of saying "a function that runs as soon as it gets defined." You can do this with anonymous functions too:
+
+```js
+(function(){
+  var x;
+  for(x = 2; x < 10; x++){
+    console.log(x + " black spiders! Ah ah ahh!")
+  }
+  console.log("I lahv counting! Ah ah ahh!")
+}());
+```
+
+This seems kind of pointless, right? The advantage is that it keeps variables from cluttering up your browser. When Javascript comes to the end of a function, it "forgets" all of the data it generated inside that function to free up memory. Keeping everything inside functions and IIFEs can have a significant impact on the performance of your website since it takes less memory to run.
+
+## Asynchronicity
```

Comment:
I think we should still talk about async, but a much easier version something like 

```js
function someAsyncFun(){
  console.log("before the setTimeout()")
  setTimeout(function(){
    console.log("this occurs in the timeout")
  }, 4000)
  console.log("after the setTimeout()")
}